### ISSUE TITLE
[Enhancement] Implement the websocket fallback as a real fallback

### ISSUE THREAD

**Issue by schlagmichdoch** [schlagmichdoch/PairDrop#228]
**What problem is solved by the new feature**
Currently, the websocket fallback (enabled via `WS_FALLBACK=true`) is only used when a device cannot use `window.RTCPeerConnection`. Some admins have expected it to be a fallback to the WebRTC connection, so it is also used whenever the WebRTC connection fails.

**Describe the feature**
If the WebRTC connection fails (3 times?) the websockets fallback is used instead.

**Additional context**
See also #215 
Only implement this after implementing traffic encryption for the websocket fallback (part of #180)

**Comment by terrytw** [comment]
First of all I would like to thank you for the quick response and also for the continuous effort on maintaining this grea project. It benefits a lot of people and we are all grateful!

1. The clients were circled in red, I remember thinking about it "a nice UI touch". There must be some other elements (intead of the change in your implementaton which I thought was the reason) which prevented the WEBRTC protocol, which leads to point 2;
2. I live in a country with heavy internet censorship, and a lot of treacherous internet situations like UDP block, multiple layer of nat, etc. Plus many public webrtc server is blocked or really slow. Webrtc is basically unusable here. 
3. I mostly used pairdrop in a local network (my home network), and I only tested websocket fallback the first time I implemented it a few months back, and it was working (looking back it was because of other reasons instead of my dummy turn server config). But the last couple of days I needed to use it in a mobile network environment where direction connection is impossible, and it fails to connect, which leads me here. 
4. I fully agree that a real fallback is the best solution. I just hope there is a way right now to establish connection without webrtc or direct connection because the real solution might be months away. But again my usecase is quite niche so I think if it is not a simple "flip of a switch" kind of deal, it won't be worth it. 

Sorry for taking up your time. Once again, really appreciate the work!

**Comment by schlagmichdoch** [comment]
There was a bug in a version some months ago that accidentally enforced all connections to be made via ws if the fallback was activated.

You can try some things:
1. Set the iceServers of the rtcconfig.json to an empty array: this could enable devices on the same network to find each other:
   ```json
   {
     "sdpSemantics": "unified-plan",
     "iceServers": []
   }
   ```
2. Use uBlock Origin (or alike) to disable WebRTC on your device
3. Use Firefox' about:config to disable WebRTC on your device
4. Enforce the WS fallback for all devices by changing this line: https://github.com/schlagmichdoch/PairDrop/blob/b36105b1cf51332bab548bc85cf3745090d706d6/public/scripts/util.js#L41
   Into this:
   ```javascript
   window.isRtcSupported = false;
   ```
   This way all clients broadcast that they do not support WebRTC.

If you worry about your privacy and try to hide your identity via VPN, step 2 or 3 would probably be wise regardless of PairDrop to prevent WebRTC leakage.

About your fourth point: I’m heavily rewriting the complete WebRTC part that was written for Snapdrop right now so I will probably get to this issue soon as well while I’m at it. I’ll add it to the pipeline for the next minor version.

**Comment by terrytw** [comment]
Massive thank you! I will definitely try step 1 and 4, they are perfect for
my need.

schlagmichdoch ***@***.***> 于 2024年2月4日周日 22:02写道：

**Comment by swiftbird07** [comment]
What is the status on this? I would love the fallback to actually work. My use case is that if I want to share between two devices where one of them is behind a corporate firewall and therefore can't use my turn server, the fallback should be used.

**Comment by schlagmichdoch** [comment]
@swiftbird07 The project came to a halt for 1-2 months because [GitHub falsely shadow hid my account and deleted loads of issues and comments](https://github.com/schlagmichdoch/PairDrop/issues/284). I have since restored everything back to its original state.

The implementation as a real fallback will be part of the next minor version `v1.11.0` which is quite big (see #267). I’m currently testing all changes and should finish in a week or two. Are you willing to test the fallback feature on a development instance?

**Comment by bt90** [comment]
@schlagmichdoch I took a look at the current codebase and apparently the websocket transfer uses base64 encoded content to be transferable as a single JSON message. Would it be possible to split this? e.g. send the control message as JSON and then send the data as a binary message?

The reason I ask is because base64 adds 33% overhead for each file.

**Comment by schlagmichdoch** [comment]
This was the easiest way to implement the websocket fallback as each message must also include the addressee to be able to relay it. Probably there are other optimized methods than sending it encoded as text. I also think all peer to peer communication should be at least transport encrypted as with WebRTC transfers but this too is not implemented yet. (See [this comment](https://github.com/schlagmichdoch/PairDrop/issues/180#issuecomment-1947523954) and the following)

I'm open to ideas but would prefer to discuss them in a separate new issue as it is not connected to this issue.

**Comment by scarab714** [comment]
Thanks for bringing up this topic. It really helped me understand the situation I was dealing with.
I was trying to use PairDrop on a company computer that's behind a web proxy, but it didn’t work, even with WS_FALLBACK=true. I initially thought there was something wrong with my setup because other app with websocket are working well so websocket wasn't blocked.

Later, after reading this topic, I tried again, this time disabling WebRTC on a Firefox browser on the other computer (not the company one), and it worked immediately.
So, I can confirm that WS_FALLBACK=true isn't a true fallback. It only works if both devices have WebRTC disabled (or blocked). If one device still has WebRTC enabled and authorized, it sticks with that, and the two devices won’t be able to discover each other.

I might say something stupid but wouldn't it be a good idea to add an button where we can manually select whether we want to switch to WS_FALLBACK mode? That way, when we know a transfer won’t work with WebRTC, we can manually force it to use WS_FALLBACK.

**Comment by thomaschampagne** [comment]
> 4. Enforce the WS fallback for all devices by changing this line: 
>      
>        
>          [PairDrop/public/scripts/util.js](https://github.com/schlagmichdoch/PairDrop/blob/b36105b1cf51332bab548bc85cf3745090d706d6/public/scripts/util.js#L41)
>             Line 41
>          in
>          [b36105b](/schlagmichdoch/PairDrop/commit/b36105b1cf51332bab548bc85cf3745090d706d6)
>               window.isRtcSupported = !!(window.RTCPeerConnection || window.mozRTCPeerConnection || window.webkitRTCPeerConnection); 
>    Into this:
>    window.isRtcSupported = false;
>    This way all clients broadcast that they do not support WebRTC.

Also used option 4 with below code adjustement to avoid forcing false for browsers which can actually handle WebRTC...

I use private browsers w/ WebRTC disabled by default:
- helium (desktop)
- vanadium on grapheneOS (android)

The old code return true even if WebRTC disabled in chromium flags of my private browsers, Below code actually make a real test with empty iceServers under a timeout race condition.

No ideal, but does the work instantly without compromises

````javascript
/**
 * Determines if WebRTC is functionally supported by attempting a connection test.
 * Uses a configurable timeout to accommodate slower networks or heavy CPU load.
 * @returns {Promise<boolean>}
 */
window.isRtcSupported = await (async () => {
    // Quick Fail: If API doesn't exist, it's definitely unsupported
    if (!(window.RTCPeerConnection || window.mozRTCPeerConnection || window.webkitRTCPeerConnection)) {
        console.log('[PairDrop] WebRTC API not found. Fallback required.');
        return false;
    }
	
    const PeerConnection = window.RTCPeerConnection || window.mozRTCPeerConnection || window.webkitRTCPeerConnection 
    const RTC_TEST_DEFAULT_TIMEOUT = 5000;
    const CONFIG_TIMEOUT = window.RTC_TEST_TIMEOUT || RTC_TEST_DEFAULT_TIMEOUT;
    
    console.log(`[PairDrop] WebRTC API found. Starting connectivity test (Timeout: ${CONFIG_TIMEOUT}ms)...`);

    const pc = new PeerConnection({ iceServers: [] }); 

    const timeoutPromise = new Promise((_, reject) => 
        setTimeout(() => {
            pc.close();
            reject(new Error(`Timeout: No candidates gathered within ${CONFIG_TIMEOUT}ms`));
        }, CONFIG_TIMEOUT)
    );

    const testPromise = new Promise((resolve, reject) => {
        let hasError = false;

        // Log specific ICE candidate errors
        pc.onicecandidateerror = (e) => {
            hasError = true;
            console.warn(`[PairDrop] WebRTC ICE Error: ${e.errorText} (Code: ${e.errorCode})`);
            
            // Detect common "UDP Blocked" signatures
            if (e.errorText.toLowerCase().includes('stun') || 
                e.errorText.toLowerCase().includes('udp') || 
                e.errorText.toLowerCase().includes('network')) {
                pc.close();
                reject(new Error('UDP Blocked by Browser Policy'));
            }
        };

        // Monitor connection state changes
        pc.oniceconnectionstatechange = () => {
            if (pc.iceConnectionState === 'failed' || pc.iceConnectionState === 'closed') {
                if (!hasError) {
                    console.log(`[PairDrop] WebRTC Connection State: ${pc.iceConnectionState}`);
                }
                if (pc.iceConnectionState === 'failed') {
                    pc.close();
                    reject(new Error('Connection Failed'));
                }
            }
        };

        // Success Condition: If we gather at least one candidate
        pc.onicecandidate = (e) => {
            if (e.candidate) {
                console.log(`[PairDrop] WebRTC Success: Gathered candidate type '${e.candidate.type}'`);
                pc.close();
                resolve(true);
            }
        };

        // Trigger candidate gathering
        pc.createDataChannel('pairdrop-test');
        pc.createOffer()
            .then(desc => pc.setLocalDescription(desc))
            .catch(err => {
                console.error('[PairDrop] WebRTC Offer Creation Failed:', err);
                reject(err);
            });
    });

    try {
        await Promise.race([testPromise, timeoutPromise]);
        console.log('[PairDrop] WebRTC Connectivity Test: PASSED');
        return true;
    } catch (err) {
        console.error(`[PairDrop] WebRTC Connectivity Test: FAILED - ${err.message}`);
        console.warn('[PairDrop] Falling back to WebSocket transport due to WebRTC failure.');
        return false;
    }
})();
````
